QUERY-SPEC.md

```diff
@@ -129,14 +102,20 @@ Then the following queries are valid:
     * -> fetches all direct-child nodes of any `dependencies` nodes in the
          document. In this case, it will fetch both `miette` and `winapi` nodes.
 
-If using an API that supports the [map operator](#map-operator), the following
-are valid queries:
-
-* `package name => val()`
-    * -> `["foo"]`.
-* `dependencies[platform] => platform`
-    * -> `["windows"]`
-* `dependencies > [] => (name(), val(), path)`
-    * -> `[("winapi", "1.0.0", "./crates/my-winapi-fork"), ("miette", "2.0.0", None)]`
-* `dependencies > [] => (name(), values(), props())`
-    * -> `[("winapi", ["1.0.0"], {"platform": "windows"}), ("miette", ["2.0.0"], {"dev": true})]`
+## Full Grammar
+
+For rules that are not defined in this grammar, see [the KDL grammar](https://github.com/kdl-org/kdl/blob/main/SPEC.md#full-grammar).
+
+```
+query := selector q-ws* "||" q-ws* query | selector
+selector := filter q-ws* selector-operator q-ws* selector | filter
+selector-operator := ">>" | ">" | "++" | "+"
+filter := matcher+
+matcher := "top()"| "()" | identifier | type | accessor-matcher
+accessor-matcher := "[" (comparison | accessor)? "]"
+comparison := accessor q-ws* matcher-operator q-ws* (type | identifier | string | number | keyword)
+accessor := "val(" number ")" | "prop(" identifier ")" | "name()" | "tag()" | "values()" | "props()" | identifier
+matcher-operator := "=" | "!=" | ">" | "<" | ">=" | "<=" | "^=" | "$=" | "*="
+
+q-ws := bom | unicode-space
+```
```

Comment:
- `identifier` should be `string`
- `val()` now also accepts decimal floats, as well as hex/octal/binary (I'm not opposed to keeping the latter)
- For consistency we might want to allow `q-ws` in the accessor and matcher brackets, at least those that take an "argument"
- `filter` now allows multiple immediately sequential `identifier`s (i.e. `string`s). Depending on how the grammar is defined with respect to backtracking etc., this would make `true` a valid query, parsed as two sequential identifiers, e.g. `tru` and `e`.

Reply:
oops. Can you take a look at the new grammar and tell me what you think?

Reply:
I think it's good to fix the order like this a bit, but we should probably mention it in the text as well. `top()` doesn't really make sense combined with other matchers (since it's not really a node) so I'm not sure I'd allow that. Type matchers are not optional at the moment, but if we fix that we need to ensure at least one matcher is used. Something like this?

```
filter := "top(" q-ws* ")" | matchers
matchers := type-matcher string? accessor-matcher* | string accessor-matcher* | accessor-matcher+
type-matcher := "(" q-ws* ")" | type
```

We can replace `q-ws` with `unicode-space` now that `bom` has been moved, but I am fine with keeping it for brevity. If we'd like `/**/` comments in KQL we could also use `ws`.

Reply:
fix'd